code/go/0chain.net/core/common/lock.go

```diff
@@ -0,0 +1,78 @@
+package common
+
+import (
+	"sync"
+)
+
+type Lock struct {
+	key string
+	// stale signals if lock is deleted from map. if true, discard use of this lock and get fresh lock
+	stale      bool
+	actualLock *sync.Mutex
+	// count how many process requires this lock
+	count   int32
+	countMu *sync.Mutex
+
+	// Parent Map
+	pMap *MapLocker
+}
+
+// Lock Acquire lock
+func (l *Lock) Lock() {
+	for {
+		l.actualLock.Lock()
+		if l.stale {
+			l = l.pMap.GetLock(l.key)
+			continue
+		}
+		break
+	}
+}
+
+func (l *Lock) Unlock() {
+	l.actualLock.Unlock()
+	l.countMu.Lock()
+	l.count--
+	if l.count == 0 {
+		l.stale = true
+		l.pMap.m.Delete(l.key)
+	}
+	l.countMu.Unlock()
+}
+
+type MapLocker struct {
+	m *sync.Map
+	// Reduce updating lock with GetLock call with same key
+	mu *sync.Mutex
+}
+
+func (m *MapLocker) GetLock(key string) (l *Lock) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	valueI, ok := m.m.Load(key)
+	if ok {
+		l = valueI.(*Lock)
+		l.countMu.Lock()
```

Comment:
Instead of using mutex to protect the l.count, we can use sync/atomic to avoid potential deadlock